File: windows-app/Plantoir.Core/Catalogs/CourseCatalog.cs
using System.Text.Json.Nodes;

namespace Plantoir.Core.Catalogs;

/// <summary>One row in a province's course catalog — what the New Course wizard's code picker searches and browses.</summary>
public sealed record CourseCatalogEntry(string Code, string ShortName, string FormalName, string Province)
{
    /// <summary>
    /// "ON:ICS3U" or "BC:MTEL-12" — unique across both provinces even where
    /// a code collides. Unused on Windows today (kept for parity with the
    /// mac's CourseCatalogEntry, where it backs SwiftUI's Identifiable for
    /// the picker's List; WinUI's AutoSuggestBox needs no item identity).
    /// </summary>
    public string Id => $"{Province}:{Code}";
}

/// <summary>
/// The two province course lists the New Course wizard's code picker offers,
/// so a teacher who doesn't already know course codes can find one by typing
/// what the course is called instead. Mirrors the mac's CourseCatalog.swift.
///
/// Reads the same ontario_secondary_courses.json and
/// british_columbia_secondary_courses.json files CourseNameCatalog reads for
/// names, loaded again here WITH which province each code belongs to —
/// CourseNameCatalog merges both files into one code-only lookup (club
/// detection and name auto-fill don't care which province a code is from)
/// and has no reason to keep the provinces apart, but a picker that offers
/// "Ontario" or "British Columbia" does.
/// </summary>
public static class CourseCatalog
{
    /// <summary>How many rows the picker offers at once — Ontario alone lists nearly 2,000 codes.</summary>
    public const int SearchResultLimit = 40;

    public static List<CourseCatalogEntry> LoadEntries(string jsonPath, string province)
    {
        var result = new List<CourseCatalogEntry>();
        if (!File.Exists(jsonPath)) return result;
        try
        {
            if (JsonNode.Parse(File.ReadAllBytes(jsonPath)) is JsonObject root)
                foreach (var (code, value) in root)
                {
                    if (value is not JsonObject entry) continue;
                    if (entry["formal_name"] is not JsonValue f || !f.TryGetValue<string>(out string? formal)) continue;
                    if (entry["short_name"] is not JsonValue s || !s.TryGetValue<string>(out string? shortName)) continue;
                    result.Add(new CourseCatalogEntry(code, shortName, formal, province));
                }
        }
        catch { }
        return result;
    }

    /// <summary>
    /// Entries from one province's catalog matching <paramref name="query"/>
    /// against code, short name, and formal name (case-insensitive
    /// substring) — codes the query PREFIXES lead the list, so typing "sch"
    /// surfaces SCH3U and SCH4U before some unrelated course whose formal
    /// name merely mentions chemistry. An empty query returns the whole
    /// catalog alphabetically by code. Capped at <paramref name="limit"/>.
    /// </summary>
    public static List<CourseCatalogEntry> Matching(IEnumerable<CourseCatalogEntry> entries, string query, int limit)
    {
        string trimmed = query.Trim();

        if (trimmed.Length == 0)
            return entries.OrderBy(e => e.Code, StringComparer.OrdinalIgnoreCase).Take(limit).ToList();

        string needle = trimmed.ToUpperInvariant();
        var codeStartsWith = new List<CourseCatalogEntry>();
        var other = new List<CourseCatalogEntry>();
        foreach (var entry in entries)
        {
            if (entry.Code.StartsWith(needle, StringComparison.OrdinalIgnoreCase)) { codeStartsWith.Add(entry); continue; }
            bool matchesElsewhere =
                entry.Code.Contains(needle, StringComparison.OrdinalIgnoreCase) ||
                entry.ShortName.Contains(needle, StringComparison.OrdinalIgnoreCase) ||
                entry.FormalName.Contains(needle, StringComparison.OrdinalIgnoreCase);
            if (matchesElsewhere) other.Add(entry);
        }
        codeStartsWith = codeStartsWith.OrderBy(e => e.Code, StringComparer.OrdinalIgnoreCase).ToList();
        other = other.OrderBy(e => e.Code, StringComparer.OrdinalIgnoreCase).ToList();

        var combined = new List<CourseCatalogEntry>(codeStartsWith);
        combined.AddRange(other);
        return combined.Take(limit).ToList();
    }
}

File: mac-app/QuartzTeachers/Catalogs/CourseCatalog.swift
import Foundation

/// One row in a province's course catalog — what the new-course wizard's
/// code picker searches and browses.
struct CourseCatalogEntry: Identifiable, Equatable {

    // MARK: - Stored properties

    let code: String
    let shortName: String
    let formalName: String

    /// "ON" or "BC".
    let province: String

    // MARK: - Computed properties

    var id: String { return province + ":" + code }

    /// Whether Plantoir ships ready-made pages for this code — the same
    /// question `ExampleContentCatalog` answers everywhere else in the
    /// wizard, asked here so the picker can mark it.
    var hasExampleContent: Bool {
        return ExampleContentCatalog.hasContent(forCode: code)
    }
}

/// The two province course lists the new-course wizard's code picker
/// offers, so a teacher who doesn't already know course codes can find
/// one by typing what the course is called instead.
///
/// Reads the same `ontario_secondary_courses.json` and
/// `british_columbia_secondary_courses.json` files `CourseNameCatalog`
/// reads for names, loaded again here WITH which province each code
/// belongs to — `CourseNameCatalog` merges both files into one
/// code-only lookup and has no reason to keep the provinces apart, but
/// a picker that offers "Ontario" or "British Columbia" does.
enum CourseCatalog {

    // MARK: - Stored properties

    static let ontarioEntries: [CourseCatalogEntry] = loadEntries(
        fileName: "ontario_secondary_courses", province: "ON"
    )
    static let britishColumbiaEntries: [CourseCatalogEntry] = loadEntries(
        fileName: "british_columbia_secondary_courses", province: "BC"
    )

    // MARK: - Functions

    static func entries(forProvince province: String) -> [CourseCatalogEntry] {
        if province == "BC" {
            return britishColumbiaEntries
        }
        return ontarioEntries
    }

    static func provinceName(forCode province: String) -> String {
        if province == "BC" {
            return "British Columbia"
        }
        return "Ontario"
    }

    /// Entries from one province's catalog matching `query` against
    /// code, short name, and formal name (case-insensitive substring) —
    /// codes the query PREFIXES lead the list, so typing "sch" surfaces
    /// SCH3U and SCH4U before some unrelated course whose formal name
    /// merely mentions chemistry. An empty query returns the whole
    /// catalog alphabetically by code, for browsing rather than
    /// searching. Capped at `limit`, since Ontario alone lists nearly
    /// 2,000 codes.
    static func matching(_ query: String, inProvince province: String, limit: Int) -> [CourseCatalogEntry] {
        let allEntries: [CourseCatalogEntry] = entries(forProvince: province)
        let trimmed: String = query.trimmingCharacters(in: .whitespaces)

        if trimmed.isEmpty {
            var sorted: [CourseCatalogEntry] = allEntries.sorted { firstEntry, secondEntry in
                return firstEntry.code < secondEntry.code
            }
            if sorted.count > limit {
                sorted = Array(sorted.prefix(limit))
            }
            return sorted
        }

        let needle: String = trimmed.uppercased()
        var codeStartsWithMatches: [CourseCatalogEntry] = []
        var otherMatches: [CourseCatalogEntry] = []
        for entry in allEntries {
            if entry.code.hasPrefix(needle) {
                codeStartsWithMatches.append(entry)
                continue
            }
            let matchesElsewhere: Bool = entry.code.contains(needle)
                || entry.shortName.uppercased().contains(needle)
                || entry.formalName.uppercased().contains(needle)
            if matchesElsewhere {
                otherMatches.append(entry)
            }
        }
        codeStartsWithMatches.sort { firstEntry, secondEntry in
            return firstEntry.code < secondEntry.code
        }
        otherMatches.sort { firstEntry, secondEntry in
            return firstEntry.code < secondEntry.code
        }

        var combined: [CourseCatalogEntry] = []
        combined.append(contentsOf: codeStartsWithMatches)
        combined.append(contentsOf: otherMatches)
        if combined.count > limit {
            combined = Array(combined.prefix(limit))
        }
        return combined
    }

    private static func loadEntries(fileName: String, province: String) -> [CourseCatalogEntry] {
        guard let url = Bundle.main.url(forResource: fileName, withExtension: "json") else {
            return []
        }
        guard let data = try? Data(contentsOf: url) else {
            return []
        }
        guard let decoded = try? JSONSerialization.jsonObject(with: data) else {
            return []
        }
        guard let dictionary = decoded as? [String: Any] else {
            return []
        }

        var result: [CourseCatalogEntry] = []
        for (code, rawEntry) in dictionary {
            guard let entry = rawEntry as? [String: Any] else {
                continue
            }
            guard let formalName = entry["formal_name"] as? String else {
                continue
            }
            guard let shortName = entry["short_name"] as? String else {
                continue
            }
            result.append(CourseCatalogEntry(
                code: code,
                shortName: shortName,
                formalName: formalName,
                province: province
            ))
        }
        return result
    }
}
